Jakarta - EE 10 - RestClientExceptionTest.testExceptionCaught failure

With the current state of `jakarta-rewrite` (for more information, see https://github.com/quarkusio/quarkus/tree/main/jakarta#build-locally), we have the following failure:

```java
[ERROR] io.quarkus.restclient.exception.RestClientExceptionTest.testExceptionCaught  Time elapsed: 0.01 s  <<< FAILURE!
org.opentest4j.AssertionFailedError: expected: <null> but was: <http://localhost/private-service>
	at org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)
	at org.junit.jupiter.api.AssertionFailureBuilder.buildAndThrow(AssertionFailureBuilder.java:132)
	at org.junit.jupiter.api.AssertNull.failNotNull(AssertNull.java:50)
	at org.junit.jupiter.api.AssertNull.assertNull(AssertNull.java:35)
	at org.junit.jupiter.api.AssertNull.assertNull(AssertNull.java:30)
	at org.junit.jupiter.api.Assertions.assertNull(Assertions.java:275)
	at io.quarkus.restclient.exception.RestClientExceptionTest.testExceptionCaught(RestClientExceptionTest.java:44)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at io.quarkus.test.QuarkusUnitTest.runExtensionMethod(QuarkusUnitTest.java:486)
	at io.quarkus.test.QuarkusUnitTest.interceptTestMethod(QuarkusUnitTest.java:400)
```

My guess is that a behavior has changed in RESTEasy but I want us to get more details about what is exactly going on before involving James Perkins.

--- a/extensions/quartz/deployment/pom.xml
+++ b/extensions/quartz/deployment/pom.xml
@@ -34,6 +34,11 @@
             <groupId>io.quarkus</groupId>
             <artifactId>quarkus-quartz</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.eclipse.transformer</groupId>
+            <artifactId>org.eclipse.transformer</artifactId>
+            <version>0.5.0</version>
+        </dependency>
 
         <dependency>
             <groupId>io.quarkus</groupId>

--- a/extensions/quartz/deployment/src/main/java/io/quarkus/quartz/deployment/JakartaEnablement.java
+++ b/extensions/quartz/deployment/src/main/java/io/quarkus/quartz/deployment/JakartaEnablement.java
@@ -1,0 +1,89 @@
+package io.quarkus.quartz.deployment;
+
+import java.nio.ByteBuffer;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+import org.eclipse.transformer.action.ActionContext;
+import org.eclipse.transformer.action.ByteData;
+import org.eclipse.transformer.action.impl.ActionContextImpl;
+import org.eclipse.transformer.action.impl.ByteDataImpl;
+import org.eclipse.transformer.action.impl.ClassActionImpl;
+import org.eclipse.transformer.action.impl.SelectionRuleImpl;
+import org.eclipse.transformer.action.impl.SignatureRuleImpl;
+import org.eclipse.transformer.util.FileUtils;
+import org.objectweb.asm.ClassReader;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.quarkus.bootstrap.classloading.QuarkusClassLoader;
+import io.quarkus.deployment.annotations.BuildProducer;
+import io.quarkus.deployment.annotations.BuildStep;
+import io.quarkus.deployment.builditem.BytecodeTransformerBuildItem;
+
+/**
+ * Quartz is compiled using references to classes in the javax packages;
+ * we need to transform these to fix compatibility with jakarta packages.
+ * We do this by leveraging the Eclipse Transformer project during Augmentation, so
+ * that end users don't need to bother.
+ */
+public class JakartaEnablement {
+
+    private static final List<String> CLASSES_NEEDING_TRANSFORMATION = List.of(
+            "org.quartz.ExecuteInJTATransaction",
+            "org.quartz.ee.servlet.QuartzInitializerServlet",
+            "org.quartz.ee.servlet.QuartzInitializerListener",
+            "org.quartz.ee.jta.JTAJobRunShell",
+            "org.quartz.ee.jta.UserTransactionHelper",
+            "org.quartz.ee.jta.UserTransactionHelper$UserTransactionWithContext",
+            "org.quartz.xml.XMLSchedulingDataProcessor",
+            "org.quartz.impl.jdbcjobstore.JTANonClusteredSemaphore");
+
+    @BuildStep
+    void transformToJakarta(BuildProducer<BytecodeTransformerBuildItem> transformers) {
+        if (QuarkusClassLoader.isClassPresentAtRuntime("jakarta.transaction.Transaction")) {
+            JakartaTransformer tr = new JakartaTransformer();
+            for (String classname : CLASSES_NEEDING_TRANSFORMATION) {
+                final BytecodeTransformerBuildItem item = new BytecodeTransformerBuildItem.Builder()
+                        .setCacheable(true)
+                        .setContinueOnFailure(false)
+                        .setClassToTransform(classname)
+                        .setClassReaderOptions(ClassReader.SKIP_DEBUG)
+                        .setInputTransformer(tr::transform)
+                        .build();
+                transformers.produce(item);
+            }
+        }
+    }
+
+    private static class JakartaTransformer {
+
+        private final Logger logger;
+        private final ActionContext ctx;
+        // We need to prevent the Eclipse Transformer to adjust the "javax" packages.
+        // Thus why we split the strings.
+        private static final Map<String, String> renames = Map.of("javax" + ".transaction", "jakarta.transaction",
+                "javax" + ".servlet", "jakarta.servlet",
+                "javax" + ".xml.bind", "jakarta.xml.bind");
+
+        JakartaTransformer() {
+            logger = LoggerFactory.getLogger("JakartaTransformer");
+            //N.B. we enable only this single transformation of package renames, not the full set of capabilities of Eclipse Transformer;
+            //this might need tailoring if the same idea gets applied to a different context.
+            ctx = new ActionContextImpl(logger,
+                    new SelectionRuleImpl(logger, Collections.emptyMap(), Collections.emptyMap()),
+                    new SignatureRuleImpl(logger, renames, null, null, null, null, null, Collections.emptyMap()));
+        }
+
+        byte[] transform(final String name, final byte[] bytes) {
+            logger.debug("Jakarta EE compatibility enhancer for Quarkus: transforming " + name);
+            final ClassActionImpl classTransformer = new ClassActionImpl(ctx);
+            final ByteBuffer input = ByteBuffer.wrap(bytes);
+            final ByteData inputData = new ByteDataImpl(name, input, FileUtils.DEFAULT_CHARSET);
+            final ByteData outputData = classTransformer.apply(inputData);
+            return outputData.buffer().array();
+        }
+    }
+
+}

--- a/extensions/resteasy-classic/resteasy-common/deployment/src/main/java/io/quarkus/resteasy/common/deployment/ResteasyCommonProcessor.java
+++ b/extensions/resteasy-classic/resteasy-common/deployment/src/main/java/io/quarkus/resteasy/common/deployment/ResteasyCommonProcessor.java
@@ -54,6 +54,7 @@ import io.quarkus.deployment.annotations.Consume;
 import io.quarkus.deployment.annotations.Record;
 import io.quarkus.deployment.builditem.CombinedIndexBuildItem;
 import io.quarkus.deployment.builditem.StaticInitConfigSourceProviderBuildItem;
+import io.quarkus.deployment.builditem.SystemPropertyBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.ReflectiveClassBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.ServiceProviderBuildItem;
 import io.quarkus.deployment.util.ServiceUtil;
@@ -152,6 +153,14 @@ public class ResteasyCommonProcessor {
                 (capabilities.isCapabilityWithPrefixPresent(Capability.RESTEASY_JSON_JACKSON)
                         // RESTEASY_JSONB or RESTEASY_JSONB_CLIENT
                         || capabilities.isCapabilityWithPrefixPresent(Capability.RESTEASY_JSON_JSONB)));
+    }
+
+    @BuildStep
+    void disableDefaultExceptionMapper(BuildProducer<SystemPropertyBuildItem> systemProperties) {
+        // RESTEasy 6.2 introduced a default exception mapper (as per JAX-RS 3.1 spec)
+        // but we want to keep Quarkus consistent and have Vert.x handle the exceptions
+        // as we have a better error handling in our Vert.x error handler.
+        systemProperties.produce(new SystemPropertyBuildItem("dev.resteasy.exception.mapper", "false"));
     }
 
     @BuildStep

--- a/independent-projects/extension-maven-plugin/pom.xml
+++ b/independent-projects/extension-maven-plugin/pom.xml
@@ -275,6 +275,7 @@
         <dependency>
             <groupId>org.apache.maven</groupId>
             <artifactId>maven-compat</artifactId>
+            <version>3.8.6</version>
             <scope>test</scope>
         </dependency>
         <dependency>

--- a/jakarta/rewrite.yml
+++ b/jakarta/rewrite.yml
@@ -72,7 +72,7 @@ recipeList:
   # JAX-RS - See below for the specs
   - org.openrewrite.maven.ChangePropertyValue:
       key: resteasy.version
-      newValue: 6.2.0.Beta1
+      newValue: 6.2.0.Final
   # JAXB - See below
   # JSON - See below
   # JSON Bind

--- a/jakarta/rewrite.yml
+++ b/jakarta/rewrite.yml
@@ -179,7 +179,7 @@ recipeList:
       newValue: 4.0.0
   - org.openrewrite.maven.ChangePropertyValue:
       key: jaxb-runtime.version
-      newValue: 4.0.0
+      newValue: 4.0.1
   # WildFly Security
   - org.openrewrite.maven.ChangePropertyValue:
       key: wildfly-elytron.version

--- a/jakarta/rewrite.yml
+++ b/jakarta/rewrite.yml
@@ -619,7 +619,7 @@ recipeList:
       newValue: 3.0.0-RC1
   - org.openrewrite.maven.ChangePropertyValue:
       key: smallrye-reactive-messaging.version
-      newValue: 4.0.0.RC1
+      newValue: 4.0.0.RC3
   - org.openrewrite.maven.ChangePropertyValue:
       key: microprofile-rest-client.version
       newValue: 3.0

--- a/jakarta/transform.sh
+++ b/jakarta/transform.sh
@@ -361,7 +361,7 @@ git cherry-pick -x ${JAKARTA_10_CDI_HASH}
 
 ## JAX-RS/RESTEasy Reactive
 git fetch origin jakarta-10-jaxrs
-git rev-list 1ac748346b91512957121e9b2f68c3b960c41565..origin/jakarta-10-jaxrs | tac | xargs git cherry-pick -x
+git rev-list 3e7b6d56854a7402f7e2cfc6edbaa49b2da0f959..origin/jakarta-10-jaxrs | tac | xargs git cherry-pick -x
 
 # Build phase
 

